clang/test/SemaSYCL/vla-kernel-param.cpp

```diff
@@ -0,0 +1,132 @@
+// RUN: %clang_cc1 -fsycl-is-device -ast-dump %s | FileCheck %s
+
+// This test checks that compiler generates correct kernel arguments for
+// VLAs.
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void a_kernel(const Func &kernelFunc) {
+  kernelFunc();
+}
+
+void foo(int *i, int x, int y) {
+  using intarray = int(*)[x]; // Motivating example
+  intarray ia = reinterpret_cast<intarray>(i);
+  a_kernel<class Canonical>([=]() {
+    ia[1][2] = 9;
+  });
+  a_kernel<class FLA>([=]() { // Not a VLA
+    int fla[4];
+    fla[0] = 9;
+  });
+  a_kernel<class SD>([=]() {
+    i[0] = 9; // single-level pointer
+  });
+  using intmdarray = int(*)[x][y];
+  intmdarray imda = reinterpret_cast<intmdarray>(i);
+  a_kernel<class MD3>([=]() {
+    imda[1][2][3] = 9; // Multi-dimensional VLA
+  });
+  using intflavlaarray = int(*)[x][4]; // VLA then a fixed-length array
+  intflavlaarray ifva = reinterpret_cast<intflavlaarray>(i);
+  a_kernel<class MDFLA>([=]() {
+    ifva[1][2][3] = 9;
+  });
+  using intvlaflaarray = int(*)[4][y]; // A fix-length array dimension followed by a VLA
+  intvlaflaarray ivfa = reinterpret_cast<intvlaflaarray>(i);
+  a_kernel<class MDFLAVLA>([=]() {
+    ivfa[1][2][3] = 9;
+  });
+}
+
+// Check Canonical parameters
+// CHECK: FunctionDecl {{.*}}Canonical{{.*}} 'void (unsigned long, int *__global *)'
+// CHECK: ParmVarDecl {{.*}} used _arg_ 'unsigned long'
+// CHECK-NEXT: ParmVarDecl {{.*}} used _arg_ 'int *__global *'
+
+// Check Canonical inits
+// CHECK-NEXT: CompoundStmt
+// CHECK-NEXT: DeclStmt
+// CHECK-NEXT: VarDecl {{.*}} cinit
+// CHECK-NEXT: InitListExpr
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'unsigned long' <LValueToRValue>
+// CHECK-NEXT: DeclRefExpr {{.*}} 'unsigned long' lvalue ParmVar {{.*}} '_arg_' 'unsigned long'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'intarray':'int (*)[x]' <AddressSpaceConversion>
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'int *__global *' <LValueToRValue>
+// CHECK-NEXT: DeclRefExpr {{.*}} 'int *__global *' lvalue ParmVar {{.*}} '_arg_' 'int *__global *'
+
+// Check FLA parameters
```

Comment:
This is a personal preference which you can disregard if you disagree. I prefer having CHECK statements next to what we are testing instead of all together at the end like this. In my opinion, it just makes it easier to read tests. Same comment for CodeGen test above

Reply:
Thanks, I will make this change.   But before I touch this file, another question.

The contents of SemaSYCL/vla-kernel-param.cpp and CodeGenSYCL/kernel-param-vla-ih.cpp are mostly the same.  The former is used for the AST dump verification and the latter for Integration Header checks.   Should I merge the two into one place?  If so, where?  Is there a preference?

Reply:
I think its better keeping it separate the way you have now.

Reply:
Please let me know what you think of the rearranged CHECKs.